Handle monitor change events via Kafka

itzg opened the pull request:
# Resolves

https://jira.rax.io/browse/SALUS-141

# What

Instead of watching etcd for agent config assignments, the ambassador now watches kafka for monitor change events.

Also introduces Introduce StringSafeJsonSerializer since resource management was expecting the attach event messages to contain the JsonSerializer java type headers. The quirk is that we still need the metrics kafka messages to not be re-serialized which is why this delegating serializer skips the JSON when it sees the incoming object is a string already.

GeorgeJahad commented:
Sorry, I don't quite understand the problem that StringSafeJsonSerializer is solving.  Under what circumstances do you get a string instead of a java object?

itzg commented:
@GeorgeJahad the "back story" is here https://github.com/racker/salus-telemetry-ambassador/pull/23  In short, the ambassador is using Avro JSON serializing for metrics, but when combining that with the regular JsonSerializer is would "double serialize" the metrics JSON into a JSON quoted string.

GeorgeJahad commented on src/test/java/com/rackspace/salus/telemetry/ambassador/services/EnvoyRegistryTest.java:
```diff
@@ -82,6 +88,14 @@ public void postsAttachEventOnAttach() throws StatusException {
+        0, 0, 0, null, 0, 0);
+    SendResult sendResult = new SendResult(null, recordMetadata);
+    ListenableFuture lf = new CompletableToListenableFutureAdapter(CompletableFuture.completedFuture(sendResult));
+    when(kafkaTemplate.send(anyString(), anyString(), any()))
```
it seems odd to add this mock now.  Why was it not needed earlier?

itzg replied:
The mock isn't new. What's new is the use of the return value from the `KafkaTemplate.send`. Since the call itself is asynchronous, but returns a completable future, I return that to the caller so that exceptions within the kafka producer thread are exposed.

itzg replied:
...it was a unit test miss from a previous change https://github.com/racker/salus-telemetry-ambassador/commit/8991d706832adf244ad635e2861bbb1f1236a1d5

GeorgeJahad commented:
sorry, just to be clear, the problem is that you are using the same serializer for two different topics.  Is that correct?

itzg commented:
> sorry, just to be clear, the problem is that you are using the same serializer for two different topics. Is that correct?

Correct. When I added the attach event publishing I introduced a bug since there were now two topics being published but needed different serialization. Ideally Kafka would let us configure serializers per topic...but it doesn't afaik :(

jjbuchan commented on src/main/java/com/rackspace/salus/telemetry/ambassador/services/MonitorEventListener.java:
```diff
@@ -0,0 +1,114 @@
+  }
+
+  public String getGroupId() throws UnknownHostException {
+    return "ambassador-"+InetAddress.getLocalHost().getHostAddress();
```
What is kubernetes setting these hosts to?

For now they must all be the same, right?  ... but in future we could configure them to be finer groups where if one goes down a new one would come up with the same name to replace it?

itzg replied:
It uses the pod names which is either 
- the deployment name with a pseudo-uuid appended for each instance of the pod
- for statefulset it will be more deterministic where it is the statefulset/deployment name with the replica number appended

jjbuchan replied:
Just realized for now we do actually want them to be unique and in future we'd want them to be more similar.  So all is still good.

jjbuchan commented on src/main/java/com/rackspace/salus/telemetry/ambassador/services/MonitorEventListener.java:
```diff
@@ -0,0 +1,114 @@
+  }
+
+  @KafkaListener(topics = "#{__listener.topic}", groupId = "#{__listener.groupId}")
+  public void handleMessage(MonitorEvent event) {
```
With this, I guess the key we set is irrelevant.

itzg replied:
Oh yeah, I guess it was. The alternative is to not use a JsonDeserializer and instead take a consumer record and deserialize the body when it's ours. I'm going to leave that option for later if we find performance data reveals a need.

jjbuchan replied:
👍

jjbuchan commented on src/main/java/com/rackspace/salus/telemetry/ambassador/services/MonitorEventListener.java:
```diff
@@ -0,0 +1,114 @@
+  @Override
+  public void onPartitionsAssigned(Map<TopicPartition, Long> assignments,
+                                   ConsumerSeekCallback callback) {
+    assignments.forEach((tp, currentOffset) -> callback.seekToEnd(tp.topic(), tp.partition()));
```
Is this jumping to the newest message on the topic?  Don't we want to consume from the last offset read?

itzg replied:
I'm not totally sure since this is the one of the race conditions I worried about when moving away from etcd as the single source of truth. Since each instance of ambassador is its own consumer group, then this partition assignment really just happens at ambassador startup. In that case, all the Envoy sessions are starting over fresh so any outstanding events would refer to envoy IDs that no longer exist.

jjbuchan replied:
Hmm, good point.

jjbuchan commented:
> Ideally Kafka would let us configure serializers per topic...but it doesn't afaik 

I had the same thought so had a quick search.  The most similar solution seems to be to use Avro for all messages, and then you can run a schema registry that kafka will use to know what schema to pull for each topic.  But that seems overkill for us now.

itzg commented:
@jjbuchan @GeorgeJahad sorry, I had some local changes I forgot to push. Can I get a small re-review?

jjbuchan commented on src/main/java/com/rackspace/salus/telemetry/ambassador/services/MetricRouter.java:
```diff
@@ -74,6 +76,7 @@ public void route(String tenantId, String envoyId,
             .setAccountType(AccountType.RCN)
             .setAccount(tenantId)
             .setTimestamp(universalTimestampFormatter.format(timestamp))
+            .setDevice(resourceId)
```
I'm not 100% convinced about this but seems fine short-term.  Really we need to know how the device will be populated consistently across all metric sources.

itzg replied:
Totally agree and I should have noted here that stuffing resource ID in here was more just a temporary strategy.
